build/platforms: Add Arty A7 platform

CBJamo opened the pull request:
Pretty straightforward, I just copied the file for the S7 and corrected the pins. Most of the info came from Digilent's XDC file, but I had to go to the schematics for the ddr interface. I doubled checked the latter, but it would be good to get another set of eyes on it.

The chipkit/shield pins are missing from here, and also from the S7, I wasn't sure if that was on purpose. The A7 also has some power monitoring stuff that isn't in here, because I wasn't sure how to name/format it.

cr1901 commented:
@CBJamo Interesting timing, as I have some Arty S7 changes I need to make but have not done so yet.

Could you split out the pmod resources into the `_connector` list, as demonstrated [here](https://github.com/enjoy-digital/litex/blob/master/litex/boards/platforms/arty.py#L111-L123)? This was an oversight on my part when initially committing Arty S7 support.

In fact, the [Arty platform file](https://github.com/enjoy-digital/litex/blob/master/litex/boards/platforms/arty.py) in LiteX is known to work, DDR and all, so check your pins/resource names against it. Ignore the `rgb_leds` resource- the one provided by LiteX uses an earlier, different layout; my method is the "current" layout.

CBJamo commented:
Ok, I checked it against the litex config. The only change was to split the ethernet into data and clocks. The xdc file has a tx, rx, and ref clocks. The litex version only has the tx and rx clocks. Not sure what's up there.

The _connector list seems like a good place for the chipkit headers. Though it's not super straightforward, they're split oddly, and the analog headers are shared between multiple pins on the FPGA. The analog header, when used as digital io, are are ck_io[14] to ck_io[25]. I'm inclined to put this in the _connectors list: 

````
("ck_io", 0, "V15 U16 P14 T11 R12 T14 T15 T16 N15 M16 V17 U18 R17 P17",
        "F5 D8 C7 E7 D7 D5 B7 B6 E6 E5 A4 A3",
        "U11 V16 M13 R10 R11 R13 R15 P15 R16 N16 N14 U17 T18 R18 P18 N17"),
````
Then handle the analog stuff seperately. The user will just have to know to not use both at the same time. I'm not sure how to handle the analog in the file.

sbourdeauducq commented on migen/build/platforms/arty_a7.py:
```diff
@@ -0,0 +1,170 @@
+        Subsignal("dq", Pins("K17", "K18", "L14", "M14")),
+        IOStandard("LVCMOS33")
+    ),
+    ("spiflash_1x", 0,  # clock needs to be accessed through STARTUPE2
```
Called spiflash/spiflash4x on other platforms.

CBJamo commented:
Fixed the spiflash naming. The s7 also needs to be changed, should I roll that into this PR?

sbourdeauducq commented:
Yes, thanks.

cr1901 commented:
@CBJamo While you are modifying S7, would you mind splitting out the `_connector` list for Arty S7 as well, like you did for Arty A7? That would save me an additional PR.

CBJamo commented:
Done and done. 

The A7 has the chipkit io (arduino shield) in the _conector list, the S7 does not. Looking at the XDC files, the S7 is missing some of the pins the A7 has, or they're shared with other functions on the board. 

How should xadc pins and power measurements be handled?

sbourdeauducq commented:
> How should xadc pins and power measurements be handled?

Your call, no one has used them before AFAIK.

The _ in spiflash_4x should also go.

CBJamo commented:
Removed the _ from both the A7 and S7.

I added xadc, modeled after how they're listed in the file for the kc705. I also changed to the same style of naming for the chipkit io. If that looks good, I'll do the same for the S7.
